mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala

```diff
@@ -317,7 +318,12 @@ final class OneVsRest @Since("1.4.0") (
     val numClasses = MetadataUtils.getNumClasses(labelSchema).fold(computeNumClasses())(identity)
     instr.logNumClasses(numClasses)
 
-    val multiclassLabeled = dataset.select($(labelCol), $(featuresCol))
+    val multiclassLabeled = getClassifier match {
+      // SPARK-21306: cache weightCol if necessary
+      case c: HasWeightCol if c.isDefined(c.weightCol) && c.getWeightCol.nonEmpty =>
+        dataset.select($(labelCol), $(featuresCol), c.getWeightCol)
+      case _ => dataset.select($(labelCol), $(featuresCol))
+    }
```

Comment:
Hi, @yanboliang . As @MLnick said, no all classifiers inherits HasWeightCol, so it might cause confusion.

In my opinion, `setWeightCol` is an attribute owned by one specific classifier, like `setProbabilityCol` and `setRawPredictionCol` for Logistic Regreesion. So I'd suggest that user should configure the classifier itself, rather than OneVsRest. Is it OK?